### ISSUE TITLE
MagicDNS doesn't register hostnames with upper-case letters

### ISSUE THREAD

**Issue by skoobasteeve** [juanfont/headscale#363]
**Bug description**

Machines are not added to MagicDNS if their hostname contains capital letters. 

**To Reproduce**

1. Set hostname on client system to something with capital letters e.g. Foo-Bar. 
2. Enroll with Headscale server using `tailscale up --login-server <URL> --auth-key <AUTH_KEY>`
3. Try to ping the client system from another client using Foo-Bar or Foo-Bar.<namespace>.<baseurl>
OR run `nslookup Foo-Bar`
4. Receive failures on both commands
5. Re-connect to the Headscale server and manually set hostname to lower-case letters using `tailscale up --hostname=foo-bar --login-server <URL>`
6. Ping and nslookup are both able to find the system at foo-bar.

**Context info**

- Headscale version: 14.0
- Tailscale client: 1.22
- Server OS: Ubuntu 20.04
- Client OS: Ubuntu 20.04

Let me know if this is a known issue or if anyone else can reproduce it. I have limited systems to test on.

Thank you!

**Comment by kyhwana** [comment]
Hmm, appears to work with tailscale ping, but not nslookup or ping in linux..

**Comment by kyhwana** [comment]
There's something funky somewhere.. Adding ``` 	hostname = strings.ToLower(hostname) ``` at line 616 in machine.go fixes this.
(As in)

``` 
} else {
		hostname = machine.Name
	}
	hostname = strings.ToLower(hostname)

	node := tailcfg.Node{

```

**Comment by restanrm** [comment]
In your opinion, should we normalize the `machine.Name` with https://github.com/juanfont/headscale/blob/b1bd17f3164c3921b5f410bcb4faa3ef4da8fba2/namespaces.go#L242-L264 or accept Uppercase letters ?

**Comment by kyhwana** [comment]
@restanrm I think it depends where the actual issue with upper case letters affecting magicDNS is? It just needs to be done before wherever they get added to magicdns dns entries. (From what I can tell)

**Comment by e-zk** [comment]
> Let me know if this is a known issue or if anyone else can reproduce it.

Not sure if it's directly related to this issue, but my Android phone with hostname "Pixel 4a" (shown via `tailscale status`), doesn't resolve with `nslookup pixel-4a` nor, `nslookup Pixel-4a`.

**Comment by ohdearaugustin** [comment]
I guess the magicdns should be conform to the RFC https://tools.ietf.org/html/rfc4343. Therefore we should accept uppercase letters in the dns name.

**Comment by kradalby** [comment]
If the issue is that uppercase is not accepted, then I think we should just lowercase it at that point, and let users have uppercase if they join with that.

I think it would be too much a nightmare to deny uppercase.

I would agree that all names should be unique in lowercase and therefor a normalise to lowercase before storing could make sense regardless. We need them to be unique

**Comment by restanrm** [comment]
@ohdearaugustin this RFC seems to have a lot of uncertainty about the behavior that we should expect (§4.2). Also if we apply this we should escape all special chars instead of replacing them (if I understand that RFC properly).

**Comment by kradalby** [comment]
A new beta is out: [v0.29.0-beta.1](https://github.com/juanfont/headscale/releases/tag/v0.29.0-beta.1).

This issue should be addressed in the release.

If the problem persists, open a new issue with full debug information: headscale config, policy, headscale logs at trace level, plus `tailscale debug netmap` and `tailscale status --json` from an affected client. See the [debugging and troubleshooting guide](https://headscale.net/development/ref/debug/) for collection steps and follow the issue template.
